src/reaction.rs

```diff
@@ -0,0 +1,428 @@
+    /// valid emoji sequences via reaction picker. On the receiver
+    /// side, abuse of the possibility to use arbitrary strings as
+    /// reactions is not different from other kinds of spam attacks
+    /// such as sending large numbers of large messages, and should be
+    /// dealt with the same way, e.g. by blocking the user.
+    fn from(reaction: &str) -> Self {
+        let mut emojis: Vec<&str> = reaction
+            .split_ascii_whitespace()
+            .filter(|&emoji| emoji.len() < 30)
+            .collect();
+        emojis.sort();
+        emojis.dedup();
+        let reaction = emojis.join(" ");
+        Self { reaction }
+    }
+}
+
+impl Reaction {
+    /// Returns true if reaction contains no emojis.
+    pub fn is_empty(&self) -> bool {
+        self.reaction.is_empty()
+    }
+
+    /// Returns a vector of emojis composing a reaction.
+    pub fn emojis(&self) -> Vec<&str> {
+        self.reaction.split(' ').collect()
+    }
+
+    /// Returns space-separated string of emojis.
+    pub fn as_str(&self) -> &str {
+        &self.reaction
+    }
+
+    /// Appends emojis from another reaction to this reaction.
+    pub fn add(&self, other: Self) -> Self {
+        let mut emojis: Vec<&str> = self.emojis();
+        emojis.append(&mut other.emojis());
+        emojis.sort();
+        emojis.dedup();
+        let reaction = emojis.join(" ");
+        Self { reaction }
+    }
+}
+
+/// Structure representing all reactions to a particular message.
+#[derive(Debug)]
+pub struct Reactions {
+    /// Map from a contact to its reaction to message.
+    reactions: BTreeMap<ContactId, Reaction>,
+}
+
+impl Reactions {
+    /// Returns vector of contacts that reacted to the message.
+    pub fn contacts(&self) -> Vec<ContactId> {
+        self.reactions.keys().copied().collect()
+    }
+
+    /// Returns reaction of a given contact to message.
+    ///
+    /// If contact did not react to message or removed the reaction,
+    /// this method returns an empty reaction.
+    pub fn get(&self, contact_id: ContactId) -> Reaction {
+        self.reactions.get(&contact_id).cloned().unwrap_or_default()
+    }
+
+    /// Returns true if the message has no reactions.
+    pub fn is_empty(&self) -> bool {
+        self.reactions.is_empty()
+    }
+}
+
+impl fmt::Display for Reactions {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut emoji_frequencies: BTreeMap<String, usize> = BTreeMap::new();
+        for reaction in self.reactions.values() {
+            for emoji in reaction.emojis() {
+                emoji_frequencies
+                    .entry(emoji.to_string())
+                    .and_modify(|x| *x += 1)
+                    .or_insert(1);
+            }
+        }
+        let mut first = true;
+        for (emoji, frequency) in emoji_frequencies {
+            if !first {
+                write!(f, " ")?;
+                first = true;
+            }
```

Comment:
the condition `!first` is never true.

i think, it should read:

```suggestion
            if !first {
                write!(f, " ")?;
            }
            first = false;
```